### PR TITLE
Ticket1750 add last update time to web dashboard

### PR DESCRIPTION
**Pull request by AdrianPotter**
### Description of work

- Added code to the webserver to get the start time for the requested instrument direct from the relevant PV
- Updated the javascript for default.html to extract and display the instrument time from the return JSON
- Refactored the block logic for making blocks from raw strings
- Added a unit test framework and unit tests for all of the above changes

### To test

https://github.com/ISISComputingGroup/IBEX/issues/1750

### Acceptance criteria

- [x] Unit tests pass
- [x] Instrument time displayed for instruments where it is available (will need to modify webserver to display localhost. See: https://github.com/ISISComputingGroup/ibex_developers_manual/wiki/Web-Dashboard
- [x] Instrument time not displayed on instruments where it is unavailable (i.e. all of them until this gets deployed)

#### Code Review
- [x] Is the code of an acceptable quality?
- [x] Have the changes been documented in the [release notes](https://github.com/ISISComputingGroup/IBEX/wiki/ReleaseNotes_Dev). If so, do they describe the changes appropriately?

### Functional Tests

- [x] Do changes function as described? Add comments below that describe the tests performed.

### Final steps
- [ ] Reviewer has moved the [release notes](https://github.com/ISISComputingGroup/IBEX/wiki/ReleaseNotes_Dev) entry for this ticket in the "Changes merged into master" section
- [ ] Deploy the new version of JSON_Bourne to the EXTWEB: https://github.com/ISISComputingGroup/ibex_developers_manual/wiki/Deployment-on-an-Instrument-Control-PC#add-instrument-to-ibex-web-dashboard

### CODE REVIEW

**Comment by DominicOram** [comment]
I'm going to be really annoying and question whether connecting to the PV directly is the best way to do this, for a couple of reasons:
* The time gathered from the PV won't actually be the time that the data was gathered from the archiver etc.
* We will have a reliance on genie python in extaweb (it currently doesn't run genie python or anything EPICS-y)

I originally suggested we use the PV as I naively thought that it was already being dumped into the archiver.

I think one of the following might be better:
* Use the time at the bottom of the archiver page
* Log the inst time into the archiver so that we can get it out like all other PVs

I'm open to debate though...

**Comment by KathrynBaker** [comment]
Add the inst time to the archiver would be my vote, it’s not difficult to add it.

**Comment by AdrianPotter** [comment]
`Adding to the archive` is equivalent to adding `info(archive, "VAL")` in the DB, correct? The PV is the devios start time for INSTETC, which is part of the underlying framework and not the regular DB file. Is there any way to add an archive flag to it without affecting all other IOCs?

Otherwise I can scrape it from the webpage without too much trouble

**Comment by AdrianPotter** [comment]
Notably, I originally considered taking the time from the currently scraped block data (hence that extra refactoring) and taking the max. The time associated with the currently collected data is the last change time though, not the last archive time. An instrument that isn't changing would report a confusing time.

I could take the max time of the last archived values. This has two drawbacks:

1. Scraping a massive amount of extra data for only a marginal benefit
1. Is that really what the user wants to see?

I would suggest that if the extaweb is already reporting a time then we would presumably want to match that.

**Comment by DominicOram** [comment]
I think scraping from the webpage seems easiest. I see no benefit in doing it any other way, if it's mildly difficult to do, as the user is not going to rely on it being 100% accurate. 

I only suggested the ticket as it would be nice to give a vague idea of when the data was valid. It's also an indication to the user that the site is being update properly.

**Comment by AdrianPotter** [comment]
No problem. Please put into rework and I'll make the modification

**Comment by KathrynBaker** [comment]
Just a quick thought, if scraping web pages anyway, would the /main page for the engine be a better option? Use the Last Written value on that page, which gives an overview of the archiver rather than any individual block?

**Comment by DominicOram** [comment]
You're right actually, I didn't realise that existed. It seems like the most trustworthy value. For reference it's at http://ndxdemo:4813/main

**Comment by AdrianPotter** [comment]
Ok, will use that

File: run_tests.py [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+    parser.add_argument('-o', '--output_dir', nargs=1, type=str, default=[DEFAULT_DIRECTORY],
+                        help='The directory to save the test reports')
+    args = parser.parse_args()
+    xml_dir = args.output_dir[0]
+
+    # Load tests from test suites
+    block_suite = unittest.TestLoader().loadTestsFromTestCase(TestBlock)
+    get_webpage_suite = unittest.TestLoader().loadTestsFromTestCase(TestGetWebpage)
+
+    print "\n\n------ BEGINNING BLOCKSERVER UNIT TESTS ------"
```
Comment (DominicOram) [inline run_tests.py]: These aren't BLOCKSERVER tests

File: get_webpage.py [hunk 2]
```diff
@@ -24,22 +24,12 @@ def shorten_title(title):
...
-    return string
-
+def _get_html_tree_from_url(url):
+    try:
+        return html.fromstring(requests.get(url).content)
+    except Exception as e:
+        logging.error("Unable to get URL tree from: " + str(url))
+        raise e
 
 def get_info(url):
```
Comment (DominicOram) [inline get_webpage.py]: PEP8 - 2 lines after method

File: get_webpage.py [hunk 3]
```diff
@@ -189,13 +137,45 @@ def scrape_webpage(host="localhost"):
...
         output["inst_pvs"] = format_blocks(get_instpvs('http://%s:%s/group?name=INST' % (host, PORT_INSTPV)))
+        output["instrument_time"] = get_instrument_time(host)
     except Exception as e:
         logging.error("Output construction failed " + str(e))
         raise e
 
     return output
 
 
+def get_instrument_time(host_instrument,get_tree=_get_html_tree_from_url):
```
Comment (DominicOram) [inline get_webpage.py]: PEP8 - missing whitespace after comma

File: get_webpage.py [hunk 4]
```diff
@@ -189,13 +137,45 @@ def scrape_webpage(host="localhost"):
...
+    """
+
+    # What we will return if we can't work out the instrument time
+    instrument_time = "Unknown"
+
+    try:
+        tree = get_tree('http://%s:%s/main' % (host_instrument, PORT_BLOCKS))
+        headers = tree.xpath("//table/tbody/tr/th")
+        info = tree.xpath("//table/tbody/tr/td")
+        assert(len(headers)==len(info))
```
Comment (DominicOram) [inline get_webpage.py]: PEP8 - missing whitespace around operator

File: get_webpage.py [hunk 5]
```diff
@@ -189,13 +137,45 @@ def scrape_webpage(host="localhost"):
...
+    except:
+        logging.error("Unable to read extweb data for instrument time for instrument: %s" % host_instrument)
```
Comment (DominicOram) [inline get_webpage.py]: This error message is the same as the one below, is it worth factoring it out?

File: test_modules/block_tests.py [hunk 6]
```diff
@@ -0,0 +1,122 @@
...
+
+    def _test_GIVEN_object_WHEN_block_from_raw_THEN_default_constructor_values_used(self, raw_object):
+        # Arrange
+        null_string = "null"
+
+        # Act
+        test_block = Block.from_raw("status", raw_object)
+
+        # Assert
+        self.assertEqual(null_string,test_block.get_alarm())
```
Comment (DominicOram) [inline test_modules/block_tests.py]: PEP8 - missing whitespace after commas

File: test_modules/block_tests.py [hunk 7]
```diff
@@ -0,0 +1,122 @@
...
+                         TestBlock._get_block_from_raw_standard_value_string().get_alarm())
+
+    def test_GIVEN_standard_format_input_WHEN_block_from_raw_THEN_returned_block_has_expected_value(self):
+        self.assertEqual(TestBlock._get_dict_of_standard_values()["value"],
+                         TestBlock._get_block_from_raw_standard_value_string().get_value())
+
+    @staticmethod
+    def _get_dict_of_dae_title_values():
+        return {"date": "2016/11/18", "time": "12:26:30.732319800",
+                "values": [40,68,65,69,32,83,73,77]}
```
Comment (DominicOram) [inline test_modules/block_tests.py]: PEP8 - missing whitespace after commas

File: test_modules/get_webpage_tests.py [hunk 8]
```diff
@@ -0,0 +1,83 @@
...
+    def test_WHEN_instrument_time_requested_THEN_instrument_time_returned(self):
+        self.assertEqual(INSTRUMENT_TIME,get_instrument_time("NDWXXXX",get_html_tree))
+
+    def test_GIVEN_instrument_WHEN_instrument_time_requested_THEN_url_contains_instrument(self):
+        instrument_name = "NDWXXXX"
+        get_instrument_time(instrument_name,get_html_tree)
+        self.assertTrue(instrument_name in URL)
+
+    def test_GIVEN_webpage_with_no_last_written_WHEN_instrument_time_requested_THEN_instrument_time_unknown(self):
+        self.assertEqual("Unknown",get_instrument_time("NDWXXXX",get_html_tree_with_no_last_written))
```
Comment (DominicOram) [inline test_modules/get_webpage_tests.py]: PEP8 - missing whitespace after comma

**Comment by DominicOram** [comment]
I'm being really picky with code formatting, apologies.

Can you also add running the unit tests to jenkins_build.bat?

Otherwise looks good :)

**Comment by AdrianPotter** [comment]
Currently there is no automated build. It's Python so the only value would be to run the unit tests. After discussion with Dom, not worth it at this stage.

**Comment by DominicOram** [comment]
Sorry, I got a little overzealous in merging this. When looking at Larmor's current configuration I get an error of:

2016-11-24 14:26:34,052 'utf8' codec can't decode byte 0xc9 in position 0: unexpected end of data

**Pull request merged**